JSON/JSON (Basic).sublime-syntax

```diff
@@ -0,0 +1,352 @@
+
+  strings:
+    - include: double-quoted-string-ahead
+
+  double-quoted-string-ahead:
+    - match: '(?=\")'
+      push: double-quoted-string
+
+  double-quoted-string:
+    - meta_scope: meta.string.json string.quoted.double.json
+    - match: '"'
+      scope: punctuation.definition.string.begin.json
+      set: inside-double-quoted-string
+
+  inside-double-quoted-string:
+    - meta_scope: meta.string.json string.quoted.double.json
+    - meta_include_prototype: false
+    - match: '"'
+      scope: punctuation.definition.string.end.json
+      pop: 1
+    - include: double-quoted-string-escape-characters
+    - match: \n
+      scope: invalid.illegal.unclosed-string.json
+      pop: 1
+
+  double-quoted-string-escape-characters:
+    - match: \\\"
+      scope: constant.character.escape.double-quote.json
+    - include: string-escape-characters
+
+  string-escape-characters:
+    - include: valid-string-escape-characters
+    - include: invalid-string-escape-characters
+
+  valid-string-escape-characters:
+    - match: \\\\
+      scope: constant.character.escape.back-slash.json
+    - match: \\\/
+      scope: constant.character.escape.forward-slash.json
+    - match: \\b
+      scope: constant.character.escape.backspace.json
+    - match: \\f
+      scope: constant.character.escape.form-feed.json
+    - match: \\n
+      scope: constant.character.escape.newline.json                             # linefeed
+    - match: \\r
+      scope: constant.character.escape.carriage-return.json
+    - match: \\t
+      scope: constant.character.escape.horizontal-tab.json
+    - match: \\u[0-9a-fA-F]{4}
+      scope: constant.character.escape.unicode-symbol.basic-multilingual-plane.json
+
+  invalid-string-escape-characters:
+    - match: \\.
+      scope: invalid.illegal.unrecognized-string-escape.json
+
+####[ Sequence ]########################################################################################################
+
+  # FIXME: trailing commas
+
+  arrays:
+    - include: empty-array
+    - match: \[
+      scope: punctuation.definition.sequence.begin.json
+      push: inside-array
+
+  empty-array:
+    - match: (\[)\s*(\])
+      scope: meta.sequence.list.empty.json
+      captures:
+        1: punctuation.definition.sequence.begin.json
+        2: punctuation.definition.sequence.end.json
+
+  inside-array:
+    - meta_scope: meta.sequence.list.json
+    - match: \]
+      scope: punctuation.definition.sequence.end.json
+      pop: 1
+    - match: (?=\S)
+      set: expect-sequence-first-item
+    # - match: '[^\s\]]'
+    #   scope: invalid.illegal.expected-any-value.json
+
+  expect-sequence-first-item:
+    - meta_scope: meta.sequence.list.json
+    - match: \]
+      scope: punctuation.definition.sequence.end.json
+      pop: 1
+    - match: ','
+      scope: invalid.illegal.expected-any-value.json
+    - match: (?=\S)
+      set: sequence-first-item
+
+  sequence-first-item:
+    - meta_scope: meta.sequence.item.first.json
```

Comment:
Do we really need `meta.sequence.item` or even `...first`? As allowed data types are quite restrictive I don't see much value of another meta scope.

We should always keep in mind ST allows limited number of scopes being stacked only. So each additional scope reduces the level of nesting being supported.

Reply:
We should try to keep the syntax as simple as possible.

Reply:
I was using `*.first`for tests with leading commas.

Reply:
In arrays and objects, the current situation is that it is:

- `meta.sequence.list` for the content scope of the array
- which is `set` replaced with `meta.sequence.item` for items in the array (or `*.first` for the first item).

So we don't stack further with arrays. Similar with objects where it's:

- `meta.mapping`, `meta.mapping.key` and `meta.mapping.value`
- not `meta.mapping meta.mapping.key` or `meta.mapping meta.mapping.value`.

Reply:
> ... replaced with `meta.sequence.item`...

Haven't read carefully enough with regards to stacking.

Not sure if an item should replace `.list` sub scope though as `meta.sequence` is meant to be very general scope which can be sub typed if needed. IMHO, the `meta.sequence.list` should cover the whole array/list.

It's the same for mappings. The whole mapping is (to be) scoped `meta.mapping` while keys and values then use `meta.mapping.key` or `meta.mapping.value`. Those special sub scopes are just appended.

But with regards to sequences it would probably problematic to have
- `meta.sequence.list.item`
- `meta.sequence.array.item`

as any plugin/color scheme needs to explicitly address all kinds of sequences in case it just wants to match sequence items in general.

Reply:
It's `meta.sequence.item` replacing `meta.sequence.list` currently. Not `meta.sequence.list.item`. I guess I should fix that.

Reply:
Understood. Since you fixed this upstream, I assume we can close this sub-thread. Thanks.